File: src/main/java/com/baulsupp/oksocial/output/DownloadHandler.java
package com.baulsupp.oksocial.output;

import com.baulsupp.oksocial.output.util.OutputUtil;
import java.io.File;
import java.io.IOException;
import okio.BufferedSource;
import okio.Okio;
import okio.Sink;

public class DownloadHandler<R> implements OutputHandler<R> {
  private ResponseExtractor<R> responseExtractor;

  private File outputFile;

  public DownloadHandler(ResponseExtractor<R> responseExtractor, File outputFile) {
    this.responseExtractor = responseExtractor;
    this.outputFile = outputFile;
  }

  @Override public void showOutput(R response) throws IOException {
    BufferedSource source = responseExtractor.source(response);

    Sink outputSink = getOutputSink(response);
    try {
      writeToSink(source, outputSink);
    } finally {
      if (!isStdout()) {
        outputSink.close();
      }
    }
  }

  public Sink getOutputSink(R response) throws IOException {
    if (isStdout()) {
      return OutputUtil.systemOut();
    } else if (outputFile.isDirectory()) {
      File responseOutputFile = new File(outputFile, responseExtractor.filename(response));
      System.err.println("Saving " + responseOutputFile);
      return Okio.sink(responseOutputFile);
    } else {
      if (outputFile.getParentFile() != null && !outputFile.getParentFile().exists()) {
        if (!outputFile.getParentFile().mkdirs()) {
          throw new IOException("unable to create directory " + outputFile);
        }
      }
      return Okio.sink(outputFile);
    }
  }

  private boolean isStdout() {
    return outputFile.getPath().equals("-");
  }

  public static void writeToSink(BufferedSource source, Sink out) throws IOException {
    while (!source.exhausted()) {
      out.write(source.buffer(), source.buffer().size());
      out.flush();
    }
  }
}

File: src/main/java/com/baulsupp/oksocial/output/DownloadHandler.kt
package com.baulsupp.oksocial.output

import com.baulsupp.oksocial.output.util.OutputUtil
import okio.BufferedSource
import okio.Okio
import okio.Sink
import java.io.File
import java.io.IOException

class DownloadHandler<R>(private val responseExtractor: ResponseExtractor<R>, private val outputFile: File) : OutputHandler<R> {

    @Throws(IOException::class)
    override fun showOutput(response: R) {
        val source = responseExtractor.source(response)

        val outputSink = getOutputSink(response)
        try {
            writeToSink(source, outputSink)
        } finally {
            if (!isStdout) {
                outputSink.close()
            }
        }
    }

    @Throws(IOException::class)
    fun getOutputSink(response: R): Sink {
        return when {
            isStdout -> OutputUtil.systemOut()
            outputFile.isDirectory -> {
                val responseOutputFile = File(outputFile, responseExtractor.filename(response))
                System.err.println("Saving " + responseOutputFile)
                Okio.sink(responseOutputFile)
            }
            else -> {
                if (outputFile.parentFile != null && !outputFile.parentFile.exists()) {
                    if (!outputFile.parentFile.mkdirs()) {
                        throw IOException("unable to create directory " + outputFile)
                    }
                }
                Okio.sink(outputFile)
            }
        }
    }

    private val isStdout: Boolean
        get() = outputFile.path == "-"

    companion object {

        @Throws(IOException::class)
        fun writeToSink(source: BufferedSource, out: Sink) {
            while (!source.exhausted()) {
                out.write(source.buffer(), source.buffer().size())
                out.flush()
            }
        }
    }
}
